RFC Enable Java 7 & 8 inspections in Idea's inspectionProfiles

paplorinc opened the pull request:
Since most modules are Java 8 now, it would make sense to reenable the related Idea inspections.
It seems to me that Idea only applies them where the language level is appropriate, see:
<img width="865" src="https://user-images.githubusercontent.com/1841944/61130610-57c4da80-a4b7-11e9-9496-10c2ae219acb.png">

Enabled a few new inspections based on what I've often seen during reviews (e.g. method may be static/final).

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
-    <inspection_tool class="SimplifyForEach" enabled="false" level="INFORMATION" enabled_by_default="false" />
-    <inspection_tool class="Since15" enabled="true" level="ERROR" enabled_by_default="true">
-      <effectiveLL value="JDK_1_8" />
-    </inspection_tool>
```
I don't think we should remove this one. Our code still needs to run on JDK8.

paplorinc replied:
This is still enabled, we're just not hard-coding the settings to Java 8, rather that it should take into consideration the language level of the module:
<img width="382" alt="Screen Shot 2019-07-15 at 22 49 51" src="https://user-images.githubusercontent.com/1841944/61248240-f456d900-a752-11e9-8ab8-c1732938ccbf.png">

paplorinc replied:
I've opened https://youtrack.jetbrains.com/issue/IDEA-218172 for Groovy & Kotlin as well

lptr replied:
I don't think this would work. Currently every project is set to language level 11 in my workspace with the default settings. We need to fix this separately (there are some cyclic dependencies that need to be fixed before IIRC?).

paplorinc replied:
For the modules, yes:
<img width="254" alt="Screen Shot 2019-07-16 at 10 31 21" src="https://user-images.githubusercontent.com/1841944/61278656-e174f080-a7b4-11e9-824a-a798199cdc61.png">

but the individual source folders seem to be correctly set to 8 or 6 - seems to me:
<img width="355" alt="Screen Shot 2019-07-16 at 10 29 43" src="https://user-images.githubusercontent.com/1841944/61278575-b68a9c80-a7b4-11e9-86b4-b12a958ebc2e.png">
<img width="304" alt="Screen Shot 2019-07-16 at 10 29 49" src="https://user-images.githubusercontent.com/1841944/61278576-b7233300-a7b4-11e9-9312-17f8a331dcc5.png">

lptr replied:
You are right. In that case I agree it's a good idea to remove this.

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
-    <inspection_tool class="Convert2Diamond" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="Convert2Lambda" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="Convert2MethodRef" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="Convert2streamapi" enabled="false" level="INFORMATION" enabled_by_default="false" />
```
I'm not completely sure about this. We should test it I think, but maybe we'll end up reverting to disabling this.

paplorinc replied:
We don't have to be sure, if we don't like it, we can always revert, as you said :)

lptr replied:
I'd rather do it as a separate PR, though, to avoid excessive warnings in our existing code.

paplorinc replied:
I've disabled `Convert2streamapi` - this is what you requested, right?

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
-    <inspection_tool class="StaticPseudoFunctionalStyleMethod" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="TryFinallyCanBeTryWithResources" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="TryWithIdenticalCatches" enabled="false" level="WARNING" enabled_by_default="false" />
+    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
```
This is not Java 8-related, right? Still seems useful.

paplorinc replied:
Yes, I've mentioned this in the description:
> Enabled a few new inspections based on what I've often seen during reviews (e.g. method may be static).

paplorinc replied:
I've opened a related issue for Groovy & Kotlin: https://youtrack.jetbrains.com/issue/KT-31561

lptr replied:
Let's move this one out of here into its own PR together with fixing all the non-final fields in that PR as well.

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
-    <inspection_tool class="TryFinallyCanBeTryWithResources" enabled="false" level="WARNING" enabled_by_default="false" />
-    <inspection_tool class="TryWithIdenticalCatches" enabled="false" level="WARNING" enabled_by_default="false" />
+    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InconsistentLanguageLevel" enabled="true" level="WARNING" enabled_by_default="true" />
```
What does this one do?

paplorinc replied:
> This inspection lists modules which depend on other modules with higher language level. Such dependencies should be broken off.

lptr replied:
This won't do anything until we actually set the right language level for each module. Not that we can't enable it.

paplorinc replied:
This isn't fully working yet, so I've disabled it.

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
-    <inspection_tool class="TryWithIdenticalCatches" enabled="false" level="WARNING" enabled_by_default="false" />
+    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InconsistentLanguageLevel" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InterfaceMayBeAnnotatedFunctional" enabled="true" level="WARNING" enabled_by_default="true" />
```
I don't think this is useful.

paplorinc replied:
K, I don't have a strong feeling about this one, removed.

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
+    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InconsistentLanguageLevel" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InterfaceMayBeAnnotatedFunctional" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="PrivateMemberAccessBetweenOuterAndInnerClass" enabled="true" level="WARNING" enabled_by_default="true" />
```
Why enable this?

paplorinc replied:
Our stack traces very often contain logs of synthetic accessors.
They aren't just noise, but before Java 11 they have performance implications as well.
It would be great if Idea could warn us about these cases.

lptr replied:
I would keep this at the default. IIRC we've never identified a case like this as a performance bottleneck. I'm worried if we enable too many warnings we'll just end up with bad case of "oh, our source files are always full of warnings, I won't even attempt to fix them."

paplorinc replied:
I have changed a few of these previously and the benchmarks have shown some improvements - or I might have misinterpreted it back then.
Are you saying the stack trace cleanup isn't worth it either?
I'll remove, and if you think it's useful I'll add it back, okay?

lptr commented on .idea/inspectionProfiles/Gradle.xml:
```diff
@@ -1,31 +1,12 @@
+    <inspection_tool class="InconsistentLanguageLevel" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="InterfaceMayBeAnnotatedFunctional" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="PrivateMemberAccessBetweenOuterAndInnerClass" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="RawUseOfParameterizedType" enabled="true" level="WARNING" enabled_by_default="true" />
```
This will make a lot of our code invalid. I think this needs to be enabled separately.

paplorinc replied:
K, removed.

paplorinc commented:
Rebased and squashed the commits to easily solve the merge conflicts.
I think we can merge it in once we agree on it, there's no point in waiting for more feedback, in my opinion: we can always come back and fix whatever we don't like.
